From 949eb8b638d213120fd5c78f7abe9e208918e72e Mon Sep 17 00:00:00 2001
From: Carlos Garnacho <carlosg@gnome.org>
Date: Tue, 2 Jul 2024 09:08:13 +0200
Subject: [PATCH] wayland: Use separate signal handlers to track pending
 activations

We cannot use a function with the same signature for signals with different
arguments, if we want to rely on the user data parameter. Separate into
two signal handlers calling the same function inside.

Fixes: b9ba34ac6f ("wayland/activation: Fix signal callback signature")
Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/3860>
---
 src/wayland/meta-wayland-activation.c | 27 ++++++++++++++++++++-------
 1 file changed, 20 insertions(+), 7 deletions(-)

diff --git a/src/wayland/meta-wayland-activation.c b/src/wayland/meta-wayland-activation.c
index e15b84fa9..67fe19c40 100644
--- a/src/wayland/meta-wayland-activation.c
+++ b/src/wayland/meta-wayland-activation.c
@@ -378,9 +378,8 @@ maybe_activate (MetaWaylandActivation *activation,
 }
 
 static void
-complete_pending_activate (MetaWindow            *window,
-                           GParamSpec            *pspec,
-                           MetaWaylandActivation *activation)
+complete_pending_activate (MetaWaylandActivation *activation,
+                           MetaWindow            *window)
 {
   g_autoptr (GPtrArray) requests = NULL;
   size_t i;
@@ -388,8 +387,7 @@ complete_pending_activate (MetaWindow            *window,
   if (!window)
     return;
 
-  g_signal_handlers_disconnect_by_func (window, complete_pending_activate,
-                                        activation);
+  g_signal_handlers_disconnect_by_data (window, activation);
 
   if (!g_hash_table_steal_extended (activation->pending_activations,
                                     window,
@@ -401,6 +399,21 @@ complete_pending_activate (MetaWindow            *window,
     maybe_activate (activation, window, requests->pdata[i]);
 }
 
+static void
+on_window_mapped_notify (MetaWindow            *window,
+                         GParamSpec            *pspec,
+                         MetaWaylandActivation *activation)
+{
+  complete_pending_activate (activation, window);
+}
+
+static void
+on_window_unmanaged (MetaWindow            *window,
+                     MetaWaylandActivation *activation)
+{
+  complete_pending_activate (activation, window);
+}
+
 static void
 add_pending_activate (MetaWaylandActivation *activation,
                       MetaWindow            *window,
@@ -419,10 +432,10 @@ add_pending_activate (MetaWaylandActivation *activation,
       requests = g_ptr_array_new_null_terminated (0, g_free, TRUE);
 
       g_signal_connect (window, "notify::mapped",
-                        G_CALLBACK (complete_pending_activate),
+                        G_CALLBACK (on_window_mapped_notify),
                         activation);
       g_signal_connect (window, "unmanaged",
-                        G_CALLBACK (complete_pending_activate),
+                        G_CALLBACK (on_window_unmanaged),
                         activation);
     }
 
-- 
2.45.2

